### PR TITLE
Support for multi-level folders

### PR DESCRIPTION
**Pull request by czardoz**

### CODE REVIEW

File: lib/converters/converter-v2-to-v1.js [hunk 1]
```diff
@@ -37,6 +37,28 @@ var _ = require('lodash').noConflict(),
...
+     *
+     * @param folderOrCollection
+     */
+    extractFolderItems: function (folderOrCollection) {
+        var i,
+            self = this,
+            folders = [],
+            items = folderOrCollection.item || folderOrCollection.items;
+
+        !_.isArray(items) && (items = [items]);
```
Comment (kunagpal) [inline lib/converters/converter-v2-to-v1.js]: Optionally, `_.castArray`(https://lodash.com/docs/4.17.4#castArray) could be used here as well.

File: lib/converters/converter-v2-to-v1.js [hunk 2]
```diff
@@ -37,6 +37,28 @@ var _ = require('lodash').noConflict(),
 _.extend(Builders.prototype, {
 
     /**
+     * Extracts all subfolders from a v2 collection or folder
+     *
+     * @param folderOrCollection
+     */
+    extractFolderItems: function (folderOrCollection) {
+        var i,
```
Comment (kunagpal) [inline lib/converters/converter-v2-to-v1.js]: This var can be avoided with a `_.forEach`

Comment (czardoz) [inline lib/converters/converter-v2-to-v1.js]: We don't use forEach in the transformer to keep it fast

File: lib/converters/converter-v2-to-v1.js [hunk 3]
```diff
@@ -333,25 +357,43 @@ _.extend(Builders.prototype, {
     },
 
     /**
+     * Creates a V1 compatible array of folder orders from a Postman collection V2
+     *
+     * @param folderOrCollection
+     */
+    folders_order: function (folderOrCollection) {
+        var itemArray = folderOrCollection.items || folderOrCollection.item,
+        allItems = _.isArray(itemArray) ? itemArray : [itemArray];
```
Comment (kunagpal) [inline lib/converters/converter-v2-to-v1.js]: _.castArray

File: lib/converters/converter-v2-to-v1.js [hunk 4]
```diff
@@ -333,25 +357,43 @@ _.extend(Builders.prototype, {
...
     folders: function (collectionV2) {
-        var self = this,
-            itemArray = collectionV2.items || collectionV2.item;
-        return _.map(_.filter(itemArray, function (element) {
-            var isFolder = (element.items || element.item);
-            return !!isFolder;
-        }), function (folder) {
+        var self = this;
+
+        var g = self.extractFolderItems(collectionV2);
```
Comment (kunagpal) [inline lib/converters/converter-v2-to-v1.js]: Can be named better

File: lib/converters/converter-v2-to-v1.js [hunk 5]
```diff
@@ -59,6 +59,28 @@ _.extend(Builders.prototype, {
     },
 
     /**
+     * Extracts all requests from a v2 collection or folder
+     *
+     * @param folderOrCollection
+     */
+    extractItems: function (folderOrCollection) {
```
Comment (vjrngn) [inline lib/converters/converter-v2-to-v1.js]: Just a thought regarding the public API. extractItems is too generic when clearly the method is focussed on requests only. Maybe rename it to convey intent?

Comment (czardoz) [inline lib/converters/converter-v2-to-v1.js]: @vjrngn This is in context of the V2 [`Items`](http://www.postmanlabs.com/postman-collection/Item.html), not any generic items.

File: lib/converters/converter-v1-to-v2.js [hunk 6]
```diff
@@ -363,6 +363,66 @@ _.extend(Builders.prototype, {
...
+        // Populate each ItemGroup with subfolders
+        _.each(collectionV1.folders, function (folderV1) {
+            var itemGroup = itemGroupCache[folderV1.id],
+                hasSubfolders = folderV1.folders_order && folderV1.folders_order.length,
+                hasRequests = folderV1.order && folderV1.order.length;
+
+            // Add subfolders
+            hasSubfolders && _.each(folderV1.folders_order, function (subFolderId) {
+                if (!itemGroupCache[subFolderId]) {
+                    // todo: figure out what to do when a collection contains a subfolder ID,
```
Comment (kunagpal) [inline lib/converters/converter-v1-to-v2.js]: Going by our previous discussions, the idea is to prune all non-existent references.

File: lib/converters/converter-v2-to-v1.js (hunk 5 above)
Comment (vjrngn) [inline lib/converters/converter-v2-to-v1.js]: Aah. ok.

File: lib/converters/converter-v1-to-v2.js [hunk 7]
```diff
@@ -414,6 +414,8 @@ _.extend(Builders.prototype, {
             });
         });
 
+        // Ideally, we could rely on collectionV1.folders_order, but since we need
```
Comment (kunagpal) [inline lib/converters/converter-v1-to-v2.js]: 👍

File: lib/converters/converter-v1-to-v2.js (hunk 6 above)
Comment (czardoz) [inline lib/converters/converter-v1-to-v2.js]: Which is what happens right now. The note is there to ensure we keep in mind that this behavior might have to change later.

**Pull request merged**